### ISSUE TITLE
Uncertainty in LightGBM

### ISSUE THREAD

**Issue by akaniklaus** [Microsoft/LightGBM#1815]
There are some techniques such as Variational Dropout (MC Dropout) to get the uncertanity of predictions (utilized in Deep Bayesian Learning). Despite the existence of Quantile Regression, I was curious if it would make sense to include them in LightGBM in the future. I would also appreciate to have a custom objective draft for Quantile Regression within the examples, as the lost function in that case is a bit more complex than the regular regression methods. Thank you very much for your time and consideration. Sincerely,

**Comment by guolinke** [comment]
@akaniklaus actually, you can achieve this by randomly drop some trees in trained GBDT model.
Another solution is like MC dropout, which you can add some noise in the input features.

**Comment by akaniklaus** [comment]
@guolinke I have an idea of ensembling several models that are hypertuned simultaneously so that they each specialize on different parts of a problem (that can be measured via their uncertainties on different input spaces). In other words, instead of a single model that would perform well in the overall problem, I want to be able to generate models that are specialized on different parts of it with a meta-layer that would ensemble their decisions considering certainty. Do you think that approach would make sense?

**Comment by guolinke** [comment]
@akaniklaus Yeah, I thank your solution can work as well.

**Comment by guolinke** [comment]
You can use shuffle_models https://lightgbm.readthedocs.io/en/latest/Python-API.html#lightgbm.Booster.shuffle_models, then use `num_iterations` in predict.

**Comment by Yatoom** [comment]
Thanks. However, I get an error saying `start_iter` is not defined. Is this a mistake in the source code for the `shuffle_models(...)` function (see `start_iteration` in the function header and `start_iter` below)?

```python
    def shuffle_models(self, start_iteration=0, end_iteration=-1):
        """Shuffle models.
        Parameters
        ----------
        start_iteration : int, optional (default=0)
            The first iteration that will be shuffled.
        end_iteration : int, optional (default=-1)
            The last iteration that will be shuffled.
            If <= 0, means the last available iteration.
        Returns
        -------
        self : Booster
            Booster with shuffled models.
        """
        _safe_call(_LIB.LGBM_BoosterShuffleModels(
            self.handle,
            ctypes.c_int(start_iter),
            ctypes.c_int(end_iter)))
```

**Comment by Yatoom** [comment]
Thanks a lot! :+1:
